tm2/pkg/std/tx_test.go

```diff
@@ -0,0 +1,216 @@
+			expectedError: "expected insufficient fee error",
+		},
+		{
+			name:          "No signatures case",
+			tx:            NewTx(msgs, NewFee(maxGasWanted, Coin{Denom: "atom", Amount: 10}), []Signature{}, "test memo"),
+			expectedError: "expected no signatures error",
+		},
+		{
+			name:          "Wrong number of signers case",
+			tx:            NewTx(msgs, NewFee(maxGasWanted, Coin{Denom: "atom", Amount: 10}), []Signature{{Signature: []byte{0x00}}, {Signature: []byte{0x01}}}, "test memo"),
+			expectedError: "expected wrong number of signers error",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			err := tc.tx.ValidateBasic()
+			if tc.expectedError == "" {
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err, tc.expectedError)
+			}
+		})
+	}
+}
+
+func Test_CountSubKeys(t *testing.T) {
+	t.Parallel()
+
+	// Single key case
+	pubKey := ed25519.GenPrivKey().PubKey()
+	require.Equal(t, 1, CountSubKeys(pubKey))
+
+	// Multi-sig case
+	pubKeys := []crypto.PubKey{ed25519.GenPrivKey().PubKey(), ed25519.GenPrivKey().PubKey()}
+	multisigPubKey := multisig.NewPubKeyMultisigThreshold(2, pubKeys)
+	require.Equal(t, len(pubKeys), CountSubKeys(multisigPubKey))
+}
+
+func Test_GetSigners(t *testing.T) {
+	t.Parallel()
+
+	addr, _ := crypto.AddressFromBech32("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5")
+	addr2, _ := crypto.AddressFromBech32("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")
+
+	testCases := []struct {
+		name     string
+		msgs     []Msg
+		expected []crypto.Address
+	}{
+		{
+			name: "Single signer case",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr,
+					msgType: "call",
+				},
+			},
+			expected: []crypto.Address{addr},
+		},
+		{
+			name: "Duplicate signers case",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr,
+					msgType: "send",
+				},
+				mockMsg{
+					caller:  addr,
+					msgType: "send",
+				},
+			},
+			expected: []crypto.Address{addr},
+		},
+		{
+			name: "Multiple unique signers case",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr,
+					msgType: "call",
+				},
+				mockMsg{
+					caller:  addr2,
+					msgType: "run",
+				},
+			},
+			expected: []crypto.Address{
+				addr,
+				addr2,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			tx := NewTx(tc.msgs, Fee{}, []Signature{}, "")
+			require.Equal(t, tc.expected, tx.GetSigners())
+		})
+	}
+}
+
+func Test_GetSignBytes(t *testing.T) {
+	t.Parallel()
+
+	msgs := []Msg{}
+	fee := NewFee(1000, Coin{Denom: "atom", Amount: 10})
+	sigs := []Signature{}
+	tx := NewTx(msgs, fee, sigs, "test memo")
+	chainID := "test-chain"
+	accountNumber := uint64(1)
+	sequence := uint64(1)
+
+	signBytes, err := tx.GetSignBytes(chainID, accountNumber, sequence)
+	require.NoError(t, err)
+	require.NotEmpty(t, signBytes)
```

Comment:
should be a golden test (match signBytes against a known example output)

Reply:
https://github.com/gnolang/gno/pull/2739/commits/5e22fe43bd1ae19e4bbdea61b24f42bffa19e6ee